python/src/deltachat/account.py

```diff
@@ -1,27 +1,157 @@
+        self._timeout = None
+
+    def __call__(self, evt_name, data1, data2):
+        self._log_event(evt_name, data1, data2)
+        self._event_queue.put((evt_name, data1, data2))
+
+    def set_timeout(self, timeout):
+        self._timeout = timeout
+
+    def get(self, timeout=None, check_error=True):
+        timeout = timeout or self._timeout
+        ev = self._event_queue.get(timeout=timeout)
+        if check_error and ev[0] == "DC_EVENT_ERROR":
+            raise ValueError("{}({!r},{!r})".format(*ev))
+        return ev
+
+    def get_matching(self, event_name_regex):
+        rex = re.compile("(?:{}).*".format(event_name_regex))
+        while 1:
+            ev = self.get()
+            if rex.match(ev[0]):
+                return ev
+
+    def _log_event(self, evt_name, data1, data2):
+        if self._debug:
+            t = threading.currentThread()
+            tname = getattr(t, "name", t)
+            print("[{}-{}] {}({!r},{!r})".format(
+                 tname, self._logid, evt_name, data1, data2))
+
+
+class Contact:
+    def __init__(self, dc_context, contact_id):
+        self.dc_context = dc_context
+        self.id = contact_id
+        # XXX do we need to free dc_contact_t? (we own it according to API)
+        self.dc_contact_t = capi.lib.dc_get_contact(self.dc_context, contact_id)
+
+    @property
+    def addr(self):
+        return ffi_unicode(capi.lib.dc_contact_get_addr(self.dc_contact_t))
+
+    @property
+    def display_name(self):
+        return ffi_unicode(capi.lib.dc_contact_get_display_name(self.dc_contact_t))
+
+    @property
+    def is_blocked(self):
+        return capi.lib.dc_contact_is_blocked(self.dc_contact_t)
+
+    @property
+    def is_verified(self):
+        return capi.lib.dc_contact_is_verified(self.dc_contact_t)
+
+
+class Chat:
+    def __init__(self, dc_context, chat_id):
+        self.dc_context = dc_context
+        self.id = chat_id
+
+    def send_text_message(self, msg):
+        """ return ID of the message in this chat.
+        'msg' should be unicode"""
+        msg = convert_bytes(msg)
+        print ("chat id", self.id)
+        return capi.lib.dc_send_text_msg(self.dc_context, self.id, msg)
 
 
 class Account:
-    def __init__(self, db_path, logcallback=None):
+    def __init__(self, db_path, _logid=None):
         self.dc_context = ctx = capi.lib.dc_context_new(
                                   capi.lib.py_dc_callback,
                                   capi.ffi.NULL, capi.ffi.NULL)
+        if hasattr(db_path, "encode"):
+            db_path = db_path.encode("utf8")
         capi.lib.dc_open(ctx, db_path, capi.ffi.NULL)
-        self._logcallback = logcallback
+        self._evhandler = EventHandler(self.dc_context)
+        self._evlogger = EventLogger(self.dc_context, _logid)
+        self._threads = IOThreads(self.dc_context)
 
     def set_config(self, **kwargs):
         for name, value in kwargs.items():
+            name = name.encode("utf8")
```

Comment:
this really smells like python2 unicode troubles going to hurt badly at some point...  i'll repeat my "please very carefully consider if python 2 support won't do more harm then good" plea...

Reply:
in python 3 kwargs always come in as unicode, and we need to encode this for the delta c-api which does not know about unicode.  This function is about config options -- we could probably just encode with "ascii" to be sure nobody tries non-ascii config names.

Reply:
otherwise i agree that moving to py3 makes sense, just not there yet becaus ei am dependent on an old project (rlcompleter2) which i wrote around 2000 and which i still heavily use ... and it does not work on python3 ...

Reply:
that being said, i think there are still quite some people i guess who have older systems and are happy if py2 bindings work.  i'd like to use the bindings server-side btw.